examples/pytorch/deepspeed-demo/train_bert_ds.py

```diff
@@ -0,0 +1,829 @@
+    # Create Datasets
+    log_dist("Creating Datasets", ranks=[0], level=logging.INFO)
+    data_iterator = create_data_iterator(
+        mask_prob=mask_prob,
+        random_replace_prob=random_replace_prob,
+        unmask_replace_prob=unmask_replace_prob,
+        tokenizer=tokenizer,
+        max_seq_length=max_seq_length,
+        batch_size=batch_size,
+    )
+    log_dist("Dataset Creation Done", ranks=[0], level=logging.INFO)
+    # Create Model
+    log_dist("Creating Model", ranks=[0], level=logging.INFO)
+    model = create_model(
+        num_layers=num_layers,
+        num_heads=num_heads,
+        ff_dim=ff_dim,
+        h_dim=h_dim,
+        dropout=dropout,
+    )
+    log_dist("Model Creation Done", ranks=[0], level=logging.INFO)
+    # DeepSpeed engine
+    log_dist("Creating DeepSpeed engine", ranks=[0], level=logging.INFO)
+    ds_config = {
+        "train_micro_batch_size_per_gpu": batch_size,
+        "optimizer": {"type": "Adam", "params": {"lr": 1e-4}},
+        "fp16": {"enabled": True},
+        "zero_optimization": {"stage": 1, "offload_optimizer": {"device": "cpu"}},
+    }
+    model, _, _, _ = deepspeed.initialize(
+        model=model, model_parameters=model.parameters(), config=ds_config
+    )
+    log_dist("DeepSpeed engine created", ranks=[0], level=logging.INFO)
+    # Load Model checkpoint
+    start_step = 1
+    if load_checkpoint_dir is not None:
+        _, client_state = model.load_checkpoint(load_dir=load_checkpoint_dir)
+        checkpoint_step = client_state["checkpoint_step"]
+        start_step = checkpoint_step + 1
+
+    # The Training Loop
+    log_dist(
+        f"Total number of model parameters: {sum([p.numel() for p in model.parameters()]):,d}",
+        ranks=[0],
+        level=logging.INFO,
+    )
+    model.train()
+    losses = []
+    for step, batch in enumerate(data_iterator, start=start_step):
+        if step >= num_iterations:
+            break
+        # Move the tensors to device
+        for key, value in batch.items():
+            batch[key] = value.to(device)
+        # Forward pass
+        loss = model(**batch)
+        # Backward pass
+        model.backward(loss)
+        # Optimizer Step
+        model.step()
+        losses.append(loss.item())
+        if step % log_every == 0:
+            log_dist(
+                "Loss: {0:.4f}".format(np.mean(losses)), ranks=[0], level=logging.INFO
+            )
+            if is_rank_0():
+                summary_writer.add_scalar("Train/loss", np.mean(losses), step)
+        if step % checkpoint_every == 0:
+            model.save_checkpoint(
+                save_dir=exp_dir, client_state={"checkpoint_step": step}
+            )
+            log_dist(
+                "Saved model to {0}".format(exp_dir), ranks=[0], level=logging.INFO
+            )
+    # Save the last checkpoint if not saved yet
+    if step % checkpoint_every != 0:
+        model.save_checkpoint(save_dir=exp_dir, client_state={"checkpoint_step": step})
```

Comment:
Will the model checkpointing be only on the rank 0 node ?

Reply:
Yes, generally, the model will be saved on shared storage (using PVC).

Reply:
@Syulin7 Do we have this check as part of `save_checkpoint()` API or we need to verify it?
Like in this [FSDP Example from PyTorch](https://pytorch.org/tutorials/intermediate/FSDP_adavnced_tutorial.html#:~:text=if%20rank%20%3D%3D%200%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20print(f%22%2D%2D%3E%20saving,)%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20torch.save(cpu_state%2C%20save_name))